tests/helpers_tests/test_condition.py

```diff
@@ -721,3 +723,594 @@ def test_audience_condition_deserializer_defaults(self):
+      )
+
+    expected_condition_log = {
+      "name": 'meters_travelled',
+      "value": 48,
+      "type": 'custom_attribute',
+      "match": 'lt'
+    }
+
+    self.assertIsNone(evaluator.evaluate(0))
+
+    mock_log = getattr(self.mock_client_logger, log_level)
+    mock_log.assert_called_once_with((
+      'Audience condition "{}" evaluated to UNKNOWN because a null value was passed '
+      'for user attribute "meters_travelled".').format(json.dumps(expected_condition_log)))
+
+  def test_substring__user_value__None(self):
+    log_level = 'debug'
+    substring_condition_list = [['headline_text', '12', 'custom_attribute', 'substring']]
+    user_attributes = {'headline_text': None}
+
+    with mock.patch('optimizely.logger.reset_logger', return_value=self.mock_client_logger):
+      evaluator = condition_helper.CustomAttributeConditionEvaluator(
+        substring_condition_list, user_attributes, self.mock_client_logger
+      )
+
+    expected_condition_log = {
+      "name": 'headline_text',
+      "value": '12',
+      "type": 'custom_attribute',
+      "match": 'substring'
+    }
+
+    self.assertIsNone(evaluator.evaluate(0))
+
+    mock_log = getattr(self.mock_client_logger, log_level)
+    mock_log.assert_called_once_with((
+      'Audience condition "{}" evaluated to UNKNOWN because a null value was '
+      'passed for user attribute "headline_text".').format(json.dumps(expected_condition_log)))
+
+  def test_exists__user_value__None(self):
+    exists_condition_list = [['input_value', None, 'custom_attribute', 'exists']]
+    user_attributes = {'input_value': None}
+
+    with mock.patch('optimizely.logger.reset_logger', return_value=self.mock_client_logger):
+      evaluator = condition_helper.CustomAttributeConditionEvaluator(
+        exists_condition_list, user_attributes, self.mock_client_logger
+      )
+
+    self.assertStrictFalse(evaluator.evaluate(0))
+
+    self.mock_client_logger.debug.assert_not_called()
+
+  def test_exact__user_value__unexpected_type(self):
+    log_level = 'warning'
+    exact_condition_list = [['favorite_constellation', 'Lacerta', 'custom_attribute', 'exact']]
+    user_attributes = {'favorite_constellation': {}}
+
+    with mock.patch('optimizely.logger.reset_logger', return_value=self.mock_client_logger):
+      evaluator = condition_helper.CustomAttributeConditionEvaluator(
+        exact_condition_list, user_attributes, self.mock_client_logger
+      )
+
+    expected_condition_log = {
+      "name": 'favorite_constellation',
+      "value": 'Lacerta',
+      "type": 'custom_attribute',
+      "match": 'exact'
+    }
+
+    self.assertIsNone(evaluator.evaluate(0))
+
+    mock_log = getattr(self.mock_client_logger, log_level)
+    mock_log.assert_called_once_with((
+      'Audience condition "{}" evaluated to UNKNOWN because a value of type "{}" was passed for '
+      'user attribute "favorite_constellation".').format(json.dumps(expected_condition_log), type({})))
```

Comment:
It'd be really nice to inline the observed type into our assertion string. Both here and below. (As long as it works for both Python 2 and Python 3... might be tricky if we want to test the `gt` and `lt` match types with string-valued user attributes.)

Reply:
We get different outputs in python 2 and 3 so can't hardcode type in string. 

> AssertionError: Expected call: warning('Audience condition "{"name": "meters_travelled", "value": 48, "type": "custom_attribute", "match": "gt"}" evaluated to UNKNOWN because a value of type "<type \'str\'>" was passed for user attribute "meters_travelled".')
> Actual call: warning('Audience condition "{"name": "meters_travelled", "value": 48, "type": "custom_attribute", "match": "gt"}" evaluated to UNKNOWN because a value of type "<class \'str\'>" was passed for user attribute "meters_travelled".')